src/heat_equation.cc

```diff
@@ -0,0 +1,585 @@
+    , dof_handler(triangulation)
+    , theta(theta_in)
+  {
+    run_name = run_name_in;
+    output_dir = output_dir_in;
+
+    use_space_adaptivity = space_adaptivity;
+    use_time_adaptivity  = time_adaptivity;
+    use_step_doubling    = step_doubling;
+
+    use_mesh = mesh;
+    cells_per_direction = cells_per_direction_in;
+
+    rhs_N = rhs_N_in;
+    rhs_sigma = rhs_sigma_in;
+    rhs_a     = rhs_a_in;
+    rhs_x0[0] = rhs_x0_x_in;
+    rhs_x0[1] = rhs_x0_y_in;
+
+    // --- Physical Parameters Initialization ---
+    density = density_in;
+    specific_heat = specific_heat_in;
+    thermal_conductivity = thermal_conductivity_in;
+    source_intensity = source_intensity_in;
+
+    time_step_tolerance = time_step_tolerance_in;
+    time_step_min       = time_step_min_in;
+
+    end_time = end_time_in;
+    time_step = initial_dt_in;
+
+    initial_global_refinement = initial_global_refinement_in;
+    n_adaptive_pre_refinement_steps = n_adaptive_pre_refinement_steps_in;
+    refine_every_n_steps = refine_every_n_steps_in;
+
+    write_vtk = write_vtk_in;
+  }
+
+  template <int dim>
+  void HeatEquation<dim>::setup_system()
+  {
+    dof_handler.distribute_dofs(fe);
+
+    std::cout << "\n===========================================\n"
+              << "[" << run_name << "] Number of active cells: " << triangulation.n_active_cells() << "\n"
+              << "[" << run_name << "] Number of degrees of freedom: " << dof_handler.n_dofs() << "\n"
+              << "===========================================\n";
+
+    constraints.clear();
+    DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+    constraints.close();
+
+    DynamicSparsityPattern dsp(dof_handler.n_dofs());
+    DoFTools::make_sparsity_pattern(dof_handler, dsp, constraints, true);
+    sparsity_pattern.copy_from(dsp);
+
+    mass_matrix.reinit(sparsity_pattern);
+    laplace_matrix.reinit(sparsity_pattern);
+    system_matrix.reinit(sparsity_pattern);
+
+    // Assemble Mass Matrix: M_ij = integral( phi_i * phi_j )
+    MatrixCreator::create_mass_matrix(dof_handler, QGauss<dim>(fe.degree + 1), mass_matrix);
+    // Multiply by Density * Specific Heat (rho * c_p)
+    mass_matrix *= (density * specific_heat);
+
+    // Assemble Laplace Matrix: A_ij = integral( grad(phi_i) * grad(phi_j) )
+    MatrixCreator::create_laplace_matrix(dof_handler, QGauss<dim>(fe.degree + 1), laplace_matrix);
+    // Multiply by Thermal Conductivity (k)
+    laplace_matrix *= thermal_conductivity;
+
+    solution.reinit(dof_handler.n_dofs());
+    old_solution.reinit(dof_handler.n_dofs());
+    system_rhs.reinit(dof_handler.n_dofs());
+  }
+
+  template <int dim>
+  void HeatEquation<dim>::solve_time_step()
+  {
+    const double rhs_norm = system_rhs.l2_norm();
+    const double tol = std::max(1e-14, 1e-8 * rhs_norm);
+
+    SolverControl            solver_control(1000, tol);
+    SolverCG<Vector<double>> cg(solver_control);
+
+    preconditioner.initialize(system_matrix, 1.0);
```

Comment:
The matrix only changes between timesteps if the timestep changes. Therefore, you might want to only initialize the preconditioner when the matrix changes, to save some computations.